Quivr-ized adam_core!

moeyensj opened the pull request:
This is a large PR, I'll highlight some areas to focus on.

moeyensj commented on adam_core/coordinates/residuals.py:
I completely reworked residuals.py as it was somewhat neglected and only in an alpha state when I ported it from THOR to adam_core. It should be significantly faster with the new batched coordinate processing rather than iterating over each coordinate. `np.einsum` also to the rescue for chi2 calculations. :)

moeyensj commented on adam_core/utils/indexable.py:
I feel like it would be appropriate to write an elegy here (not eulogy, there is no praise in this thing's existence) here. ChatGPT?

spenczar replied:
In the vast, interconnected web of Python code, there resided a digital entity of such intricate design, an artifact of human endeavor known as 'Indexable'. It made its dwelling within the confines of a script titled 'indexable.py', a realm wherein it diligently strived to fulfill its purpose. Today, we gather here to honor the cessation of its existence, to commemorate the solemn demise of this labyrinthine Python class, 'Indexable'.

Indexable was conceived with a breadth of capabilities, each designed to enable a certain kind of interaction, a particular relationship between data structures. Its iteration was a steady process, dutifully recommencing at the familiar origin point of zero. This was Indexable's testament to tradition, a steady adherence to the beginning, an eternal loop of purposeful effort.

Among the manifold operations it performed, the 'yield_chunks' method carried a particular elegance. It fragmented and offered pieces of itself with a voluntary surrender, with the simplicity of a humble servant. It was an exercise in division, a testament to the balance that Indexable sought to imbue within the code.

However, it was in the arena of sorting values that the complexity of Indexable was truly exposed. It handled arrays, masked and unmasked, times, and attributes of varied natures. It was a balancing act of a dizzying magnitude. In its complexity, however, it bore a burden that often shrouded its potential. It had to navigate its own intricate mechanisms, often erring on the side of inefficiency, its algorithms winding like an ancient labyrinth.

The task of concatenating indexables, likewise, brought forth its unwieldy nature. The process was akin to trying to unite the disparate branches of an old tree, each limb twisting in its own direction, each leaf whispering its own tale. This attempt at unity, while admirable in intention, was unfortunately a slow and laborious process.

Despite these flaws, one must extend a courteous nod to Indexable's attempt to adapt and mold itself according to the demands of the task at hand. It protected unsliceable data structures, ensuring a semblance of order amidst the chaos. But alas, these noble intentions were marred by the tedious inefficiency that pervaded its operations.

And now, as we mark the end of Indexable, it is perhaps necessary to remember it with a sense of detached understanding. Yes, Indexable was flawed, its mechanisms unwieldy, its speed less than desirable. Yet it represented an effort to navigate complexity, to bring order within a myriad of data types and operations.

In its passing, Indexable leaves behind a legacy of lessons, a testament to the constant search for balance between complexity and efficiency. It serves as a gentle reminder of our shared purpose in this realm of code - to simplify, to streamline, and above all, to serve with humility.

Indexable, despite its limitations, endeavored to be of service. Now, as it fades into the annals of digital history, we honor its memory. In the silent echoes of its code, in the legacy it leaves behind, Indexable continues to live on. Its spirit remains a beacon, guiding us towards the pursuit of elegance and efficiency in our code. It is a solemn adieu to Indexable, our complicated companion in the realm of Python. In its memory, we strive to design better, to code smarter, and to remember that even in the most convoluted corners of code, simplicity and efficiency must be our guiding lights.

moeyensj commented on adam_core/orbits/orbits.py:
Cartesian coordinates are now the default coordinate type for Orbits. Note that I override the `.to_dataframe` and `.from_dataframe` methods in order to flatten the covariance matrices and allow sigmas to be stored in dataframes too. Each coordinate class has this behavior supported (check io.py for details).

moeyensj commented:
@spenczar Do you think we should up the minimum Python version for adam_core to 3.11? Or is it worth making quivr available on less older versions?

moeyensj commented on adam_core/coordinates/conversions.py:
I've kept these unit conversion function even though Coordinates no longer have an attached property that defines them. We will want to think about how our unit conversions are going to work in the future, but for the time being working in defaults is good enough I think.

moeyensj commented on adam_core/coordinates/covariances.py:
```diff
@@ -27,49 +29,141 @@
+        """
+        return cls.from_matrix(sigmas_to_covariances(sigmas))
+
+    def to_dataframe(
```
Allows covariances to be flattened in data frames, which is nice because then they can be serialized to CSV fairly trivially.

moeyensj commented on adam_core/coordinates/covariances.py:
```diff
@@ -27,49 +29,141 @@
+        return cls.from_arrays([cov])
+
+    @classmethod
+    def from_sigmas(cls, sigmas: np.ndarray) -> "CoordinateCovariances":
```
Sometimes you really just don't want to define all the elements in a covariance matrix and instead just create one from the 1-sigma uncertainties.

moeyensj commented on adam_core/coordinates/keplerian.py:
```diff
@@ -31,263 +37,72 @@
-        Periapsis distance
-        """
-        return self.a * (1 - self.e)
+        return self.a.to_numpy() * (1 - self.e.to_numpy())
```
@spenczar Here is an example I think where it could be nice for the getters in `quivr` to return numpy arrays, or have some method to do that. This property will return the periapsis distance as an array, but if I want to access semi-major axis it's returned as a pyarrow object.

spenczar commented on adam_core/coordinates/cometary.py:
```diff
@@ -407,31 +200,39 @@ def to_cartesian(self) -> CartesianCoordinates:
             )
         else:
-            covariances_cartesian = None
+            covariances_cartesian = np.zeros(
```
```suggestion
            covariances_cartesian = np.empty(
```

moeyensj replied:
Fixed for all coordinate representations.

spenczar commented on adam_core/coordinates/cometary.py:
```diff
@@ -491,43 +298,77 @@ def from_cartesian(
 
         return coords
 
+    def to_cometary(self) -> "CometaryCoordinates":
+        return self
+
+    @classmethod
+    def from_cometary(
+        cls, cometary_coordinates: "CometaryCoordinates"
+    ) -> "CometaryCoordinates":
+        return cometary_coordinates
+
```
These look deletable. Maybe this is for a type checker though?

moeyensj replied:
Deleted.

spenczar commented on adam_core/coordinates/cometary.py:
```diff
@@ -491,43 +298,77 @@ def from_cartesian(
+            DataFrame containing coordinates.
+        """
+        return coords_to_dataframe(
+            self, list(COMETARY_COLS.keys()), sigmas=sigmas, covariances=covariances
```
Same comment here as for CartesianCoordinates' dataframes methods and the COLS thing.

spenczar reviewed (commented):
Partial review - I have read 
- adam_core.coordinates.cartesian
- adam_core.coordinates.cometary
- adam_core.coordinates.conversions
- adam_core.coordinates.covariances
- adam_core.coordinates.frame
- adam_core.coordinates.origin

moeyensj commented on adam_core/coordinates/cartesian.py:
```diff
@@ -24,448 +32,140 @@
-            units=units,
-        )
-        return
+    x = Float64Field(nullable=True)
+    y = Float64Field(nullable=True)
+    z = Float64Field(nullable=True)
+    vx = Float64Field(nullable=True)
+    vy = Float64Field(nullable=True)
+    vz = Float64Field(nullable=True)
+    times = Times.as_field(nullable=True)
+    covariances = CoordinateCovariances.as_field(nullable=True)
+    origin = Origin.as_field(nullable=False)
+    frame = Frame.as_field(nullable=False)
```
This is definitely more of a me thing, I'm still building the mental model so I was finding the explicitness helpful, especially on inspection of the source code to understand the fields of the class. Happy to remove it.

Maybe `quivr` could have a "summarize" method or something that describes the assumptions for the fields? I bet pyarrow already has something like that or actually I guess I could look at the schema.

moeyensj replied:
I was assuming fields were not nullable as well.

akoumjian replied:
I kind of prefer explicitly nullable unless this is the default arrow behavior we are mimicking? Also, why do we want the position and the velocity to be nullable? I'm guessing there is some case where this object is useful without some of those values.

moeyensj replied:
> Also, why do we want the position and the velocity to be nullable? I'm guessing there is some case where this object is useful without some of those values.

I'd like to have support for missing values. A single optical observation is lon, lat with rho, vrho, vlon, vlat missing. A radar observation can give all 6. This is extended to all coordinate classes, though I could see us removing that constraint for Keplerian and Cometary elements.

spenczar commented on adam_core/utils/tests/test_benchmarks.py:
```diff
@@ -1,52 +0,0 @@
-from .test_indexable import TestIndexable, concatenate
-
-
-@pytest.mark.parametrize(
```
It might be nice to benchmark some of the new stuff. Shouldn't be in this revision though.

spenczar commented on setup.cfg:
```diff
@@ -15,6 +15,7 @@ install_requires =
     pandas
     requests
     scipy
+    quivr
```
```suggestion
    quivr>=0.3.2
```

spenczar reviewed (commented):
Bit more

spenczar commented on adam_core/coordinates/keplerian.py:
```diff
@@ -31,263 +37,72 @@
-        Periapsis distance
-        """
-        return self.a * (1 - self.e)
+        return self.a.to_numpy() * (1 - self.e.to_numpy())
```
Indeed, that's pretty strange feeling. I'll have to keep thinking about ways to make this nicer.

moeyensj commented on adam_core/coordinates/times.py:
```diff
@@ -0,0 +1,21 @@
+
+    # TODO: @spenczar - We could have this class store times
+    # as two integers... :)
+    mjd = Float64Field(nullable=True)
+
+    # Could replace this with an enum of some kind
+    scale = StringField(nullable=True)
```
If we make these fields not nullable, can I still make `Times.as_field(nullable=True`)? Just want to make that sure we don't force coordinates to have times.

spenczar replied:
Correct, the whole struct can be nullable; my proposal would just effectively be that if you have an mjd, you need to also have a scale.

moeyensj commented on adam_core/coordinates/cartesian.py:
```diff
@@ -24,448 +32,140 @@
-        self._values[:, 0] = np.nan
-        self._values[:, 0].mask = True
+    def values(self) -> np.ndarray:
+        return self.table.to_pandas()[["x", "y", "z", "vx", "vy", "vz"]].values
```
Very nice! This is the method I was looking for.

moeyensj commented on adam_core/coordinates/tests/test_cartesian.py:
```diff
@@ -1,30 +1,33 @@
-    covariances with an identity matrix should return the same coordinates and covariances.
-    """
+    # Test rotation of Cartesian coordinates. Rotating coordinates and their
+    # covariances with an identity matrix should return the same coordinates and covariances.
```
I'll write ones for the cardinal axes. Rotate X-axis to Y-axis, etc..

moeyensj commented on adam_core/coordinates/cartesian.py:
```diff
@@ -480,169 +180,182 @@ def rotate(self, matrix: np.ndarray, frame_out: str) -> "CartesianCoordinates":
-                "coordinates that already have the desired origin."
-            )
-            logger.info(info)
+    @classmethod
+    def from_cartesian(
+        cls, cartesian: "CartesianCoordinates"
+    ) -> "CartesianCoordinates":
+        return cartesian
```
I was thinking that all 4 representations should have as many of the same methods as possible. Of course, this is violated with `rotate` and `translate` attached only to Cartesian coordinates. I can remove these.

moeyensj commented on adam_core/coordinates/cartesian.py:
```diff
@@ -480,169 +180,182 @@ def rotate(self, matrix: np.ndarray, frame_out: str) -> "CartesianCoordinates":
-        coords_rotated = deepcopy(np.ma.dot(self._values, matrix.T))
-        coords_rotated[self._values.mask] = np.NaN
+        # Extract coordinate values into a masked array and mask NaNss
+        masked_coords = np.ma.masked_array(self.values, fill_value=np.NaN)
```
Let's hold off on that until we think we might need it. With how `calc_residuals` is now written we have effectively negated the need for masked arrays.

moeyensj commented on adam_core/orbits/orbits.py:
```diff
@@ -1,196 +1,65 @@
-            self._object_ids = np.array(["None" for i in range(len(coordinates))])
+    orbit_ids = StringField(nullable=True)
+    object_ids = StringField(nullable=True)
+    coordinates = CartesianCoordinates.as_field(nullable=True)
```
Completely agree!

moeyensj commented on adam_core/coordinates/keplerian.py:
```diff
@@ -34,260 +33,71 @@
+class KeplerianCoordinates(Table):
 
-        return
+    a = Float64Field(nullable=True)
+    e = Float64Field(nullable=True)
+    i = Float64Field(nullable=True)
+    raan = Float64Field(nullable=True)
+    ap = Float64Field(nullable=True)
+    M = Float64Field(nullable=True)
+    times = Times.as_field(nullable=True)
```
Let me propose the following: let's make Cometary and Keplerian elements not nullable (let's be conservative), then if we run into a case in the future where allowing nulls is something we need we can modify it then.

moeyensj commented on adam_core/coordinates/io.py:
```diff
@@ -0,0 +1,127 @@
+    cls: "Type[CoordinateType]", df: pd.DataFrame, coord_names: List[str]
+) -> "CoordinateType":
+    """
+    Return coordinates from a pandas DataFrame.
```
Fixed.

moeyensj commented on adam_core/coordinates/spherical.py:
```diff
@@ -1,13 +1,16 @@
+from .times import Times
+
+config.update("jax_enable_x64", True)
+config.update("jax_platform_name", "cpu")
```
Removed, it's only transform.py now.

moeyensj commented on adam_core/coordinates/spherical.py:
```diff
@@ -27,283 +38,86 @@
+    def values(self) -> np.ndarray:
+        return self.table.to_pandas()[
+            ["rho", "lon", "lat", "vrho", "vlon", "vlat"]
+        ].values
```
Added in [7eede77](https://github.com/B612-Asteroid-Institute/adam_core/pull/17/commits/7eede7722d4cc1352503b432395c92d36950e813). Forgot to add this one to the batch, but I modified it during the rebase for all coordinate representations.

moeyensj commented on adam_core/coordinates/covariances.py:
```diff
@@ -27,49 +29,141 @@
-        sigmas_ = sigmas
+    D = sigmas.shape[1]
+    identity = np.identity(D, dtype=sigmas.dtype)
+    covariances = np.einsum("kj,ji->kij", sigmas**2, identity, order="C")
```
Blood sacrifices were made in the making of this code

moeyensj commented on adam_core/coordinates/frame.py:
```diff
@@ -0,0 +1,17 @@
+from quivr import StringField, Table
+
+
+# TODO: Replace with DictionaryField or similar
+#       Investigate whether this class is even necessary
```
I think we agree that the reason to put them in a class of their own is if we want to attach some custom methods to it, which for the time being we don't have yet. 

What I see us doing, especially as we add more frames like a Galactic frame (needed for visual stuff), is attaching a rotation matrix to each Frame class that rotates it to ecliptic J2000? Kind of like how we define transformations to and from cartesian. Our rotation matrices currently live in constants.py but we could move them to frames.py. I'd need to think a bit about the interface. So maybe this could work for a follow-up PR?

moeyensj commented on adam_core/coordinates/io.py:
```diff
@@ -0,0 +1,127 @@
+            coord_names=coord_names,
+            sigmas=sigmas,
+        )
+        if not covariances:
+            cov_cols = [col for col in df_cov.columns if "cov_" in col]
+            df_cov = df_cov.drop(columns=cov_cols)
```
This behaves as intended, at least in my testing. If `covariances and sigmas` then the sigmas get inserted before the covariances but they will both be present. If only sigmas, then sigmas are inserted but no covariances. If only covariances, then covariances are inserted but not sigmas. If neither, none of the uncertainties are inserted. 

```
orbits.to_dataframe(sigmas=False, covariances=False)
```
orbit_ids | object_ids | times.jd1 | times.jd2 | times.scale | x | y | z | vx | vy | vz | origin.code | frame.name
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
None | (2013 RR165) | 2456982.0 | -0.5 | tdb | -0.410387 | 2.682955 | 0.057486 | -0.010122 | -0.00084 | -0.000421 | SUN | ecliptic

```
orbits.to_dataframe(sigmas=True, covariances=False)
```
orbit_ids | object_ids | times.jd1 | times.jd2 | times.scale | x | y | z | vx | vy | vz | origin.code | frame.name | sigma_x | sigma_y | sigma_z | sigma_vx | sigma_vy | sigma_vz
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
None | (2013 RR165) | 2456982.0 | -0.5 | tdb | -0.410387 | 2.682955 | 0.057486 | -0.010122 | -0.00084 | -0.000421 | SUN | ecliptic | 8.341792e-07 | 2.766898e-07 | 6.477729e-07 | 1.158898e-09 | 1.857893e-09 | 2.000738e-09

```
orbits.to_dataframe(sigmas=False, covariances=True)
```
orbit_ids | object_ids | times.jd1 | times.jd2 | times.scale | x | y | z | vx | vy | vz | origin.code | frame.name | cov_x_x | cov_y_x | cov_y_y | cov_z_x | cov_z_y | cov_z_z | cov_vx_x | cov_vx_y | cov_vx_z | cov_vx_vx | cov_vy_x | cov_vy_y | cov_vy_z | cov_vy_vx | cov_vy_vy | cov_vz_x | cov_vz_y | cov_vz_z | cov_vz_vx | cov_vz_vy | cov_vz_vz
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
None | (2013 RR165) | 2456982.0 | -0.5 | tdb | -0.410387 | 2.682955 | 0.057486 | -0.010122 | -0.00084 | -0.000421 | SUN | ecliptic | 6.958550e-13 | 3.407623e-14 | 7.655723e-14 | 1.127173e-14 | -8.208209e-15 | 4.196098e-13 | -4.869665e-16 | 2.458330e-16 | 1.037337e-17 | 1.343045e-18 | 1.359690e-15 | 2.266828e-16 | 7.405595e-17 | -3.436471e-19 | 3.451767e-18 | 2.265675e-17 | 3.350965e-17 | -5.808007e-16 | -9.023744e-20 | -5.216186e-20 | 4.002951e-18

```
orbits.to_dataframe(sigmas=True, covariances=True)
````
orbit_ids | object_ids | times.jd1 | times.jd2 | times.scale | x | y | z | vx | vy | vz | origin.code | frame.name | sigma_x | sigma_y | sigma_z | sigma_vx | sigma_vy | sigma_vz | cov_x_x | cov_y_x | cov_y_y | cov_z_x | cov_z_y | cov_z_z | cov_vx_x | cov_vx_y | cov_vx_z | cov_vx_vx | cov_vy_x | cov_vy_y | cov_vy_z | cov_vy_vx | cov_vy_vy | cov_vz_x | cov_vz_y | cov_vz_z | cov_vz_vx | cov_vz_vy | cov_vz_vz
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
None | (2013 RR165) | 2456982.0 | -0.5 | tdb | -0.410387 | 2.682955 | 0.057486 | -0.010122 | -0.00084 | -0.000421 | SUN | ecliptic | 8.341792e-07 | 2.766898e-07 | 6.477729e-07 | 1.158898e-09 | 1.857893e-09 | 2.000738e-09 | 6.958550e-13 | 3.407623e-14 | 7.655723e-14 | 1.127173e-14 | -8.208209e-15 | 4.196098e-13 | -4.869665e-16 | 2.458330e-16 | 1.037337e-17 | 1.343045e-18 | 1.359690e-15 | 2.266828e-16 | 7.405595e-17 | -3.436471e-19 | 3.451767e-18 | 2.265675e-17 | 3.350965e-17 | -5.808007e-16 | -9.023744e-20 | -5.216186e-20 | 4.002951e-18

moeyensj replied:
Uhm.. this looks so much better in the "preview" view then it does once you submit the comment.

moeyensj commented on adam_core/coordinates/origin.py:
```diff
@@ -0,0 +1,106 @@
+#       Investigate whether this class is even necessary
+class Origin(Table):
+
+    code = StringField(nullable=False)
```
Similar to `Frames`, we do end up actually building in some functionality for gravitational parameter. I think I'll keep it as is for now. This will need some thought for observatories. Hoping to start that work tomorrow.

moeyensj commented on adam_core/coordinates/io.py:
```diff
@@ -0,0 +1,127 @@
+    """
+    # Gather times and put them into a dataframe
+    df_times = coords.times.to_dataframe(flatten=True)
+    if len(df_times.columns) > 0:
```
Yeah, I think you are correct. I'll remove it.

moeyensj commented on adam_core/coordinates/tests/test_origin.py:
```diff
@@ -0,0 +1,27 @@
+def test_origin_eq__raises():
+    # Test that an error is raised when an unsupported type is passed
+    origin = Origin.from_kwargs(code=["SUN", "EARTH", "MARS"])
+    with pytest.raises(NotImplementedError):
```
Fixed, it is now a TypeError.

moeyensj commented:
@spenczar I've addressed most of your comments so far. Two things I will spawn off into two separate PRs:
- [ ] Unit tests for coordinate transformations
- [ ] Unit tests for cartesian translations and rotations

@akoumjian Is it okay for me to add those as JIRA tickets so we track them in the backlog? 

Last items that need review:
- [ ] adam_core/coordinates/residuals.py
- [ ] adam_core/coordinates/transforms.py
- [ ] adam_core/orbits/orbits.py
- [ ] adam_core/coordinates/keplerian.py**
- [ ] adam_core/coordinates/cometary.py**

** These ones I honestly consider optional since they are so similar to the other representations in terms of function and structure. The comments Spencer made on Cartesian and Spherical coordinates have already been applied to these two. 

There are a couple of discussion items that I've left unresolved so that we can keep those around for future reference.

moeyensj commented on adam_core/coordinates/residuals.py:
```diff
@@ -1,161 +1,212 @@
+        )
+
+    N, D = observed.values.shape
+    p = np.zeros(N, dtype=np.float64)
+    chi2s = np.zeros(N, dtype=np.float64)
```
I've always defaulted to initializing arrays that I will fill with some values with `np.zeros`. More habit than intent, I just find that `np.empty` is not as tidy:
```
import numpy as np

a = np.empty((10))
print(a)
```
```
array([4.66583234e-316, 0.00000000e+000, 6.94459595e-310, 5.69163624e-321,
       5.69163624e-321, 0.00000000e+000, 0.00000000e+000, 5.04214149e-316,
       5.81912810e-316, 6.94459834e-310])
```

What do you suggest? Should I be using `np.empty` for filling and not `np.zeros`?

spenczar replied:
Right, `empty` doesn't initialize with zero values, it just gives raw memory. But this isn't notebook code, so these intermediate arrays are never going to get printed, so I don't think that concern really is important. I'd use empty when you're really just trying to allocate a slab of space for filling in later (like here)

moeyensj commented on adam_core/coordinates/residuals.py:
```diff
@@ -1,161 +1,212 @@
+    return batch_indices, batch_dimensions, batch_coords, batch_covariances
+
+
+def calculate_residuals(
+    observed: CoordinateType, predicted: CoordinateType
+) -> Residuals:
```
Great idea.

moeyensj replied:
`Residuals.calculate(observed, predicted)?`

spenczar commented on adam_core/coordinates/residuals.py:
```diff
@@ -1,161 +1,212 @@
+        batch_covariances,
+    ) = batch_coords_and_covariances(observed.values, observed.covariances.to_matrix())
+
+    for i, (indices, dimensions, coords, covariances) in enumerate(
+        zip(batch_indices, batch_dimensions, batch_coords, batch_covariances)
+    ):
+        if not np.all(np.isnan(covariances)):
+            # Calculate the chi2 for each coordinate
+            chi2_values = calculate_chi2(residuals[indices], covariances)
+
+            # Calculate the probability for each coordinate
+            p[indices] = 1 - chi2.cdf(np.sqrt(chi2_values), dof[indices])
+
+            # Set the chi2 for each coordinate
+            chi2s[indices] = chi2_values
+
+        else:
+            # If the covariance matrix is all NaNs, then the chi2 is NaN
+            chi2s[indices] = np.nan
+            p[indices] = np.nan
```
`i` appears unused; do you need the enumerate?

moeyensj replied:
Removed.

spenczar reviewed (commented):
intermediate comments

spenczar commented on adam_core/coordinates/tests/test_residuals.py:
```diff
@@ -0,0 +1,194 @@
+    # the chi2 calculation is correct.
+    residuals = np.array(
+        [
+            [1, 1, 1],
+            [1, 1, 1],
```
It looks like perhaps later tests do this? Your call

moeyensj replied:
Done: https://github.com/B612-Asteroid-Institute/adam_core/blob/quivr-coords/adam_core/coordinates/tests/test_residuals.py#L7/

spenczar commented on adam_core/coordinates/transform.py:
```diff
@@ -1434,13 +1424,28 @@ def transform_coordinates(
-        if set_rho_nan:
-            coords_out._values[:, 0] = COORD_FILL_VALUE
-            coords_out._values[:, 0].mask = 1
 
-        if set_vrho_nan:
-            coords_out._values[:, 3] = COORD_FILL_VALUE
-            coords_out._values[:, 3].mask = 1
+        if set_rho_nan or set_vrho_nan:
+            values_out = coords_out.values
+
+            if set_rho_nan:
+                values_out[:, 0] = np.NaN
+
+            if set_vrho_nan:
+                values_out[:, 3] = np.NaN
+
+            coords = SphericalCoordinates.from_kwargs(
+                rho=values_out[:, 0],
+                lat=values_out[:, 1],
+                lon=values_out[:, 2],
+                vrho=values_out[:, 3],
+                vlat=values_out[:, 4],
+                vlon=values_out[:, 5],
+                covariances=coords.covariances,
```
The covariances aren't getting masked with nulls here.

This whole behavior - implicitly casting to a unit sphere - seems error prone, to me. If you converted SphericalCoordinates to Keplerian, for example, you'd get an `a` and a `p` etc that are dependent on this imputed value, and I'm not sure what the covariances would mean at all.

While it's a little bit more convenient to automatically do this, the implications seem too dramatic. I think this makes me _more_ certain that users should call some sort of `coords.to_unit_sphere(self)` explicitly if they want this behavior, rather than putting it inside of this function.

moeyensj replied:
I agree and I like your suggestion. I've added a `.to_unit_sphere()`.

spenczar commented on adam_core/coordinates/transform.py:
```diff
@@ -1306,14 +1310,14 @@ def transform_coordinates(
             raise ValueError("".join(frame_err).format("frame_out"))
     else:
-        frame_out = coords.frame
+        frame_out = str(coords.frame.name[0])
 
-    origin_err = ["{} should be one of:\n", "'heliocenter' or 'barycenter'"]
+    origin_err = ["{} should be one of:\n", "'SUN' or 'SOLAR_SYSTEM_BARYCENTER'"]
     if origin_out is not None:
-        if origin_out != "barycenter" and origin_out != "heliocenter":
+        if origin_out != "SOLAR_SYSTEM_BARYCENTER" and origin_out != "SUN":
             raise ValueError("".join(origin_err).format("origin_out"))
```
Why is this error getting generated with string formatting, instead of being a literal?

spenczar reviewed (commented):
I'd like to propose a pretty different overall structure for `transform_coordinates`. Instead of accepting a string which maps to a class, just accept the class itself as an argument.

If you do that, a lot of the big switches get much smaller. I made some other changes (removing the unit-sphere conversion, and switching to using enums directly, and ended up with this:
```py
def transform_coordinates(
    coords: CoordinateType,
    representation_out: type[*_CoordinateClasses],
    frame_out: Optional[Literal["ecliptic", "equatorial"]] = None,
    origin_out: Optional[OriginCodes] = None,
) -> CoordinateType:
    """
    Transform coordinates between frames ('ecliptic', 'equatorial')
    and/or representations ('cartesian', 'spherical', 'keplerian').

    Parameters
    ----------
    coords : `~adam_core.coordinates.Coordinates`
        Coordinates to transform between representations and frames.
    representation_out : `~adam_core.coordinates.Coordinates`
        Desired coordinate type or representation of the output coordinates.
    frame_out : {'ecliptic', 'equatorial'}
        Desired reference frame of the output coordinates.
    origin_out : {'SUN', 'SOLAR_SYSTEM_BARYCENTER'}
        Desired origin of the output coordinates.
    Returns
    -------
    coords_out : `~adam_core.coordinates.Coordinates`
        Coordinates in desired output representation and frame.

    Raises
    ------
    ValueError
        If frame_in, frame_out are not one of 'equatorial', 'ecliptic'.
        If representation_in, representation_out are not one of 'cartesian',
            'spherical', 'keplerian', 'cometary'.

    TypeError
        If coords is not a `~adam_core.coordinates.Coordinates` object.
        If origin_out is not a `~adam_core.coordinates.OriginCodes` object.
        """
    # Check that coords is a thor.coordinates.Coordinates object
    if not isinstance(coords, _CoordinateClasses,):
        raise TypeError("unsupported coordinate type: {}".format(type(coords)))

    if frame_out not in {None, "equatorial", "ecliptic"}:
        raise ValueError("unsupported frame_out: {}".format(frame_out))

    if origin_out is not None and not isinstance(origin_out, OriginCodes):
        raise TypeError("unsupported origin_out type: {}".format(type(origin_out)))

    if representation_out not in _CoordinateClasses:
        raise ValueError(
            "Unsupported representation_out: {}".format(representation_out)
        )

    coord_frame = str(coords.frame.name[0])
    coord_origin = str(coords.origin.code[0])
    if frame_out is None:
        frame_out = coord_frame

    if origin_out is None:
        origin_out = coord_origin

    if type(coords) == representation_out:
        if coord_frame == frame_out and coord_origin == origin_out:
            return coords
        
    cartesian = coords.to_cartesian()

    # Translate coordinates to new origin (if different from current)
    if cartesian.origin != origin_out:
        cartesian = cartesian_to_origin(cartesian, origin_out)

    # Rotate coordinates to new frame (if different from current)
    if cartesian.frame != frame_out:
        cartesian = cartesian_to_frame(cartesian, frame_out)

    return representation_out.from_cartesian(cartesian)
```

spenczar reviewed (commented):
OK! I think I've reviewed everything.

moeyensj commented on adam_core/coordinates/transform.py:
```diff
@@ -1331,11 +1335,7 @@ def transform_coordinates(
-        and origin_out is None
-        and np.all(coords.origin == origin_out)
-    ):
+    if coords.frame == frame_out and origin_out is None and coords.origin == origin_out:
```
Refactor of transform_coordinates fixed this issue.

moeyensj commented on adam_core/coordinates/transform.py:
```diff
@@ -1306,14 +1310,14 @@ def transform_coordinates(
         if frame_out != "equatorial" and frame_out != "ecliptic":
             raise ValueError("".join(frame_err).format("frame_out"))
     else:
```
Frame is now a StringAttribute on coordinates. :+1:

moeyensj commented on adam_core/coordinates/transform.py:
```diff
@@ -1237,25 +1241,25 @@ def cartesian_to_frame(
-    origin_out : {'heliocenter', 'barycenter'}
+    origin_out : {'SUN', 'SOLAR_SYSTEM_BARYCENTER'}
         Desired origin of the output coordinates.
     unit_sphere : bool
```
Done! I've add a to_unit_sphere for spherical coordinates.

moeyensj commented:
> I'd like to propose a pretty different overall structure for `transform_coordinates`. Instead of accepting a string which maps to a class, just accept the class itself as an argument.
> 
> If you do that, a lot of the big switches get much smaller. I made some other changes (removing the unit-sphere conversion, and switching to using enums directly, and ended up with this:
> 
> ```python
> def transform_coordinates(
>     coords: CoordinateType,
>     representation_out: type[*_CoordinateClasses],
>     frame_out: Optional[Literal["ecliptic", "equatorial"]] = None,
>     origin_out: Optional[OriginCodes] = None,
> ) -> CoordinateType:
>     """
>     Transform coordinates between frames ('ecliptic', 'equatorial')
>     and/or representations ('cartesian', 'spherical', 'keplerian').
> 
>     Parameters
>     ----------
>     coords : `~adam_core.coordinates.Coordinates`
>         Coordinates to transform between representations and frames.
>     representation_out : `~adam_core.coordinates.Coordinates`
>         Desired coordinate type or representation of the output coordinates.
>     frame_out : {'ecliptic', 'equatorial'}
>         Desired reference frame of the output coordinates.
>     origin_out : {'SUN', 'SOLAR_SYSTEM_BARYCENTER'}
>         Desired origin of the output coordinates.
>     Returns
>     -------
>     coords_out : `~adam_core.coordinates.Coordinates`
>         Coordinates in desired output representation and frame.
> 
>     Raises
>     ------
>     ValueError
>         If frame_in, frame_out are not one of 'equatorial', 'ecliptic'.
>         If representation_in, representation_out are not one of 'cartesian',
>             'spherical', 'keplerian', 'cometary'.
> 
>     TypeError
>         If coords is not a `~adam_core.coordinates.Coordinates` object.
>         If origin_out is not a `~adam_core.coordinates.OriginCodes` object.
>         """
>     # Check that coords is a thor.coordinates.Coordinates object
>     if not isinstance(coords, _CoordinateClasses,):
>         raise TypeError("unsupported coordinate type: {}".format(type(coords)))
> 
>     if frame_out not in {None, "equatorial", "ecliptic"}:
>         raise ValueError("unsupported frame_out: {}".format(frame_out))
> 
>     if origin_out is not None and not isinstance(origin_out, OriginCodes):
>         raise TypeError("unsupported origin_out type: {}".format(type(origin_out)))
> 
>     if representation_out not in _CoordinateClasses:
>         raise ValueError(
>             "Unsupported representation_out: {}".format(representation_out)
>         )
> 
>     coord_frame = str(coords.frame.name[0])
>     coord_origin = str(coords.origin.code[0])
>     if frame_out is None:
>         frame_out = coord_frame
> 
>     if origin_out is None:
>         origin_out = coord_origin
> 
>     if type(coords) == representation_out:
>         if coord_frame == frame_out and coord_origin == origin_out:
>             return coords
>         
>     cartesian = coords.to_cartesian()
> 
>     # Translate coordinates to new origin (if different from current)
>     if cartesian.origin != origin_out:
>         cartesian = cartesian_to_origin(cartesian, origin_out)
> 
>     # Rotate coordinates to new frame (if different from current)
>     if cartesian.frame != frame_out:
>         cartesian = cartesian_to_frame(cartesian, frame_out)
> 
>     return representation_out.from_cartesian(cartesian)
> ```

I've added this refactor as part of the PR. However, I think I'll need to see how I end up using this function to really understand what it should be doing to a greater degree. It's not obvious if the origin transformation is something I'll do often through here rather than hitting a lower-level API. I'll know a lot more once this starts getting used!

moeyensj commented on adam_core/orbits/query/horizons.py:
```diff
@@ -156,20 +154,27 @@ def query_horizons(
+            Time(vectors["datetime_jd"].values, scale="tdb", format="jd")
+        )
+        origin = Origin.from_kwargs(code=["SUN" for i in range(len(times))])
+        frame = Frame.from_kwargs(name=["ecliptic" for i in range(len(times))])
```
Frame is now a StringAttribute, but I've kept Origin as a class. I think this is the right call especially as we add observatory codes.

moeyensj commented on adam_core/coordinates/transform.py:
```diff
@@ -1237,25 +1241,25 @@ def cartesian_to_frame(
+    coords: CoordinateType,
     representation_out: str,
     frame_out: Optional[str] = None,
     origin_out: Optional[str] = None,
```
Done!
